url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/VoronoiNeighbors.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * 
+ */
+package utility.geo;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Scanner;
+
+
+import mesh.Delaunay;
+
+import utility.Compare.buildBtreeOfPoints;
+import utility.security.DataIO;
+
+/**
+ * @author chenqian
+ *
+ */
+public class VoronoiNeighbors {
+
+	private ArrayList<float[]> points = null;
+	private ArrayList<Integer> idmap = null;
+	private Delaunay delaunay = null;
+	
+	public void generateVoronoiNeighbors(String sourceFileName, String destFileName){
+		String line = null;
+		LineNumberReader lr = null;
+		try {
+			lr = new LineNumberReader(new FileReader(sourceFileName));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileReader*. Currently, there are execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception. Ensure the object returned by *FileReader()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).